dialogue-apache-hc5-client/src/test/java/com/palantir/dialogue/hc5/ApacheHttpClientBlockingChannelTest.java

```diff
@@ -0,0 +1,122 @@
+import com.palantir.dialogue.HttpMethod;
+import com.palantir.dialogue.PathTemplate;
+import com.palantir.dialogue.Request;
+import com.palantir.dialogue.UrlBuilder;
+import com.palantir.dialogue.core.BaseUrl;
+import java.net.URI;
+import java.net.URL;
+import java.util.Map;
+import org.apache.hc.core5.http.ClassicHttpRequest;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.CsvSource;
+import org.junit.jupiter.params.provider.ValueSource;
+
+class ApacheHttpClientBlockingChannelTest {
+
+    @ParameterizedTest
+    @ValueSource(strings = {"GET", "PUT", "POST"})
+    void createRequest(String method) throws Exception {
+
+        BaseUrl baseUrl = BaseUrl.of(new URL("https://www.example.com/foo/api"));
+        Endpoint endpoint = new Endpoint() {
+            private final PathTemplate pathTemplate = PathTemplate.builder()
+                    .fixed("fixed")
+                    .variable("endpoint")
+                    .variable("index")
+                    .build();
+
+            @Override
+            public HttpMethod httpMethod() {
+                return HttpMethod.valueOf(method);
+            }
+
+            @Override
+            public String serviceName() {
+                return "testService";
+            }
+
+            @Override
+            public String endpointName() {
+                return "testEndpoint";
+            }
+
+            @Override
+            public String version() {
+                return "1.2.3";
+            }
+
+            @Override
+            public void renderPath(ListMultimap<String, String> params, UrlBuilder url) {
+                pathTemplate.fill(params, url);
+            }
+        };
+
+        Request request = Request.builder()
+                .pathParams(Map.of("fixed", "bar", "endpoint", "baz", "index", "42"))
+                .putQueryParams("q", "1")
+                .putQueryParams("test", "true")
+                .build();
+
+        ClassicHttpRequest httpRequest = ApacheHttpClientBlockingChannel.createRequest(baseUrl, endpoint, request);
+        assertThat(httpRequest.getMethod()).isEqualTo(method);
+        assertThat(httpRequest.getPath())
+                .isEqualTo(httpRequest.getRequestUri())
+                .isEqualTo("/foo/api/fixed/baz/42?q=1&test=true");
+        assertThat(httpRequest.getUri())
+                .isEqualTo(URI.create("https://www.example.com/foo/api/fixed/baz/42?q=1&test=true"));
+    }
+
+    @ParameterizedTest
+    @CsvSource({
+        "http://example.com, example.com, -1,",
+        "https://www.example.com, www.example.com, -1,",
+        "https://www.example.com:443, www.example.com, 443,",
+        "https://www.example.com/path/to/foo/bar, www.example.com, -1,",
+        "https://www.example.com/path/to/foo/bar?baz=quux&hello=world#hash-octothorpe, www.example.com, -1,",
+        "https://user@www.example.com:8443/path/to/foo/bar?baz=quux&hello=world#hash-octothorpe ,"
```

Comment:
Let's add a couple test cases for encoding edge cases:

* A URL which includes percent-escaped values in the users password: `https://user:slash%2Fslash@www.example.com`
* A URL which includes bracket-formatted ipv6 address: `https://user@[::1]/path` (perhaps with and without a port)

Reply:
Thanks, added some more test coverage.